web-components/src/components/coachmark/Coachmark.ts

```diff
@@ -0,0 +1,135 @@
+  @property({ type: String }) placement: coachmarkPlacement = "auto";
+  @property({ type: Boolean }) show = false;
+  @property({ type: String }) color = "default";
+
+  @query(".md-coachmark__popper") popper!: HTMLDivElement;
+  @query(".md-coachmark__reference") reference!: HTMLDivElement;
+
+  private slotContent: Element[] | null = null;
+
+  connectedCallback() {
+    super.connectedCallback();
+  }
+
+  notifyCoachCreate() {
+    this.dispatchEvent(
+      new CustomEvent("coach-create", {
+        bubbles: true,
+        composed: true,
+        detail: {
+          placement: this.placement,
+          reference: this.reference,
+          popper: this.popper
+        }
+      })
+    );
+  }
+
+  coachAction() {
+    this.dispatchEvent(
+      new CustomEvent("coach-action", {
+        bubbles: true,
+        composed: true,
+        detail: {
+          popper: this.popper
+        }
+      })
+    );
+  }
+
+  handleSlotChange(event: Event) {
+    const slot = event.target as HTMLSlotElement;
+    if (slot) {
+      const slotContent = slot.assignedElements({ flatten: true });
+      if (slotContent.length) {
+        this.slotContent = slotContent;
+      }
+    }
+  }
+
+  protected update(changedProperties: PropertyValues) {
+    super.update(changedProperties);
+    if (changedProperties.has("show")) {
+      if (this.show) {
+        this.setFocusableElements!();
+        if (this.focusableElements && this.focusableElements.length) {
+          this.focusableElements[0].focus();
+        }
+      }
+    }
+  }
+
+  get coachWrapClassMap() {
+    return {
+      "md-coachmark--active": !!this.show
+    };
+  }
+
+  get coachClassMap() {
+    return {
+      [`md-coachmark__popper--${this.placement}`]: !!this.placement,
+      [`md-coachmark__popper--${this.color}`]: !!this.color
+    };
+  }
+
+  render() {
+    return html`
+      <div class="md-coachmark ${classMap(this.coachWrapClassMap)}">
+        <div class="md-coachmark__popper ${classMap(this.coachClassMap)}" tabindex="0">
+          <div class="md-coachmark__content" >
+            ${this.message
+              ? this.message
+              : html`
+                  <slot name="coachmark-content" @slotchange=${this.handleSlotChange}></slot>
+                `}
+            <div class="md-coachmark__action">
+              ${this.hidebutton 
+                ? html`<slot name="coachmark-action"></slot>` 
+                : html`<md-button @button-click=${this.coachAction}>${this.actionname}</md-button>`}
+            </div>
+          </div>
+          <div id="arrow" class="md-coachmark__arrow" data-popper-arrow></div>
+        </div>
+        <div
+          class="md-coachmark__reference"
+          @click=${() => this.notifyCoachCreate()}
+          aria-describedby="coachmark"
+        >
+          <slot></slot>
+        </div>
+      </div>
+    `;
+  }
+
+  static get styles() {
+    return [reset, styles];
+  }
+}
```

Comment:
Nit: I think it works either way, but this style loader is usually placed just before the render method in most other components, and that seems like a nice location for it to consistently be.

Reply:
Agree. Fixed.